### ISSUE TITLE
Mean Hausdorff metric/loss function

### ISSUE THREAD

**Issue by Dale-Black** [Project-MONAI/MONAI#1039]
I am working on a hybrid Hausdorff loss function and I would love to contribute to this project since it has been such a big help for me. I think a Hausdorff metric and possibly loss function would be really useful to add to this repo. I am inexperienced with contributing to open source so let me know if this is something anyone is interested in

**Comment by Nic-Ma** [comment]
Hi @Dale-Black ,

Thanks for your interest here! Welcome to contribute a PR, we just merged another PR for Hausdorff metrics.
@wyli , did you researcher guys also discuss Hausdorff loss function before? Any concern here?

Thanks.

**Comment by wyli** [comment]
Hi @Dale-Black thanks for your interest here. Is this loss function based on some research papers, would be great to have more technical details. also:
- A basic Hausdorff metric is implemented (https://github.com/Project-MONAI/MONAI/blob/72df264c8ee34ef307262227bca27805dbbd92b4/monai/metrics/hausdorff_distance.py)
- please follow these contributing steps: https://github.com/Project-MONAI/MONAI/blob/master/CONTRIBUTING.md#preparing-pull-requests we can definitely help prepare the PR

**Comment by Dale-Black** [comment]
Hello @wyli thank you for getting back to me. This Hausdorff loss function will be based off of these two papers (https://arxiv.org/pdf/1806.07564.pdf and https://arxiv.org/pdf/1904.10030.pdf). It would require a mean Hausdorff metric which is similar to but much more computationally expensive than the Hausdorff metric that was merged with MONAI yesterday. I will get started on the PR for the mean Hausdorff metric and go from there.

**Comment by rijobro** [comment]
Is the mean Hausdorff distance computed by looping over each of the coordinates in X and finding the minimum distance to any of the points in Y, and taking the average of that? If so, you  should be able to reuse everything up until the [edge detection](https://github.com/Project-MONAI/MONAI/blob/master/monai/metrics/hausdorff_distance.py#L78), and then do loops across `coords_1` and `coords_2`. But you're right, it does sound like it would be slow. 

There's a possibility that someone else will be modifying similar code to you (see [this](https://github.com/Project-MONAI/MONAI/issues/889#issuecomment-693247632) thread), hopefully that won't result in any git conflicts.

**Comment by Dale-Black** [comment]
@rijobro Yes it would look just like the already implemented Hausdorff metric with the exception of the loops like you were describing. Here is an example

    def compute_mean_hausdorff_distance(
        seg_1: Union[np.ndarray, torch.Tensor],
        seg_2: Union[np.ndarray, torch.Tensor],
        label_idx: int,
        directed: bool = False,) -> float:

        """

        Compute the mean Hausdorff distance. 

        Args:
            seg_1: first binary or labelfield image.
            seg_2: second binary or labelfield image.
            label_idx: for labelfield images, convert to binary with
                `seg_1 = seg_1 == label_idx`.
            directed: calculate directed Hausdorff distance (defaults to `False`).
        """
        if torch.is_tensor(seg_1):
            seg_1 = seg_1.detach().cpu().numpy()
        if torch.is_tensor(seg_2):
            seg_2 = seg_2.detach().cpu().numpy()

        if seg_1.size == 0 or seg_1.shape != seg_2.shape:
            raise ValueError("Labelfields should be the same shape (and non-zero number of elements")

        # If not binary images, convert them
        if seg_1.dtype != bool:
            seg_1 = seg_1 == label_idx
        if seg_2.dtype != bool:
            seg_2 = seg_2 == label_idx

        # Check both have at least 1 voxel with desired index
        if not (np.any(seg_1) and np.any(seg_2)):
            raise ValueError(f"Labelfields should have at least 1 voxel containing the desired labelfield, {label_idx}")

        # Do binary dilation and use XOR to get edges
        edges_1 = binary_dilation(seg_1) ^ seg_1
        edges_2 = binary_dilation(seg_2) ^ seg_2

        # Extract coordinates of these edge points
        coords_1 = np.argwhere(edges_1)
        coords_2 = np.argwhere(edges_2)

        min_euc_list_u = []
        min_euc_list_v = []

        for i in range(coords_1.shape[0]):
            euc_list = []
            for j in range(coords_2.shape[0]):
                euc = euclidean(coords_1[i,:], coords_2[j,:])
                euc_list.append(euc)
            min_euc_list_u.append(min(euc_list))

        for i in range(coords_2.shape[0]):
            euc_list = []
            for j in range(coords_1.shape[0]):
                euc = euclidean(coords_2[i,:], coords_1[j,:])
                euc_list.append(euc)
            min_euc_list_v.append(min(euc_list))

        return max(np.array(min_euc_list_u).mean(), np.array(min_euc_list_v).mean())



This is so slow though that it might be more practical to first implement this using cython. Any suggestions are much appreciated

**Comment by rijobro** [comment]
From eq. 4 of the first article you linked (see attached screenshot), it looks you should be returning the sum, not the max of the two components:

```
return np.array(min_euc_list_u).mean() + np.array(min_euc_list_v).mean()
```

![image](https://user-images.githubusercontent.com/33289025/93434662-b9a57f00-f8c8-11ea-8f43-cb3656d2e834.png)


As for speed, yeah, I guess that cuda, numba or cython are going to be your friend. I don't have much experience with these unfortunately. @ericspod has been looking into this a bit recently, I believe.

**Comment by ericspod** [comment]
[This](https://link.springer.com/chapter/10.1007%2FBFb0015093) has the undirected Hausdorff defined as the max of the directed.

**Comment by rijobro** [comment]
But that's for the Hausdorff distance, not the **mean** Hausdorff distance, right?

**Comment by imadtoubal** [comment]
Were there any developments with regard to this? I have a working Hausdorff loss implementation (albeit a little slow) and would love to contribute. (My implementation is based on: https://arxiv.org/pdf/1904.10030.pdf, code: https://github.com/PatRyg99/HausdorffLoss)
